src/Renderer/Label2DRenderer.js

```diff
@@ -0,0 +1,183 @@
+                    return;
+                }
+            }
+        }
+
+        for (let i = minx; i <= maxx; i++) {
+            for (let j = miny; j <= maxy; j++) {
+                this.grid[i][j] = true;
+            }
+        }
+
+        this.visible.push(obj);
+    }
+}
+
+const viewProjectionMatrix = new THREE.Matrix4();
+const vector = new THREE.Vector3();
+
+/**
+ * This renderer is inspired by the
+ * [`THREE.CSS2DRenderer`](https://threejs.org/docs/#examples/en/renderers/CSS2DRenderer).
+ * It is instanciated in `c3DEngine`, as another renderer to handles Labels.
+ */
+class Label2DRenderer {
+    constructor(infoTileLayer) {
+        this.domElement = document.createElement('div');
+        this.domElement.style.overflow = 'hidden';
+        this.domElement.style.position = 'absolute';
+        this.domElement.style.top = 0;
+        this.domElement.style.zIndex = 1;
+
+        // Used to destroy labels that are not added to the DOM
+        this.garbage = document.createElement('div');
+        this.garbage.style.display = 'none';
+        this.domElement.appendChild(this.garbage);
+
+        this.halfWidth = 0;
+        this.halfHeight = 0;
+
+        this.grid = new ScreenGrid();
+
+        this.infoTileLayer = infoTileLayer;
+    }
+
+    setSize(width, height) {
+        this.domElement.style.width = `${width}`;
+        this.domElement.style.height = `${height}`;
+
+        this.halfWidth = width / 2;
+        this.halfHeight = height / 2;
+
+        this.grid.width = width;
+        this.grid.height = height;
+        this.grid.x = Math.ceil(width / 30);
+        this.grid.y = Math.ceil(height / 30);
+
+        this.grid.resize();
+    }
+
+    render(scene, camera) {
+        this.grid.reset();
+
+        viewProjectionMatrix.multiplyMatrices(camera.projectionMatrix, camera.matrixWorldInverse);
+
+        this.culling(scene, this.infoTileLayer.currentMaxTileZoom, this.infoTileLayer.displayed.extent);
+
+        this.grid.visible.forEach(label => label.move());
+        this.grid.hidden.forEach((label) => { label.visible = false; });
+    }
+
+    culling(object, currentMaxZoom, extent) {
+        if (!object.isLabel) {
+            if (!object.visible) {
+                this.hideNodeDOM(object);
+                return;
+            }
+
+            // Don't go further if the node can't be in the screen space (it
+            // does not need to be rendered to continue the culling)
+            if (object.extent && !object.extent.intersectsExtent(extent)) {
+                this.hideNodeDOM(object);
+                return;
+            }
+
+            this.showNodeDOM(object);
+
+            object.children.forEach(c => this.culling(c, currentMaxZoom, extent));
+        // By verifying the maxzoom and the presence of the label inside the
+        // visible extent, we can filter more labels.
+        } else if (object.maxzoom <= currentMaxZoom || !extent.isPointInside(object.coordinates)) {
```

Comment:
the object visibility and material visibility aren't enough to filter out the labels?
may be check `tile.material.visible` and if the children tile have label.

Reply:
No, it isn't: a material not visible does not mean a label is not visible anymore.